clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -163,127 +159,180 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
         Map<String, List<TopicPartition>> assignment = new HashMap<>(
-            consumerToOwnedPartitions.keySet().stream().collect(Collectors.toMap(c -> c, c -> new ArrayList<>(minQuota))));
+            consumerToOwnedPartitions.keySet().stream().collect(Collectors.toMap(c -> c, c -> new ArrayList<>(maxQuota))));
 
+        List<TopicPartition> assignedPartitions = new ArrayList<>();
         // Reassign as many previously owned partitions as possible
         for (Map.Entry<String, List<TopicPartition>> consumerEntry : consumerToOwnedPartitions.entrySet()) {
             String consumer = consumerEntry.getKey();
             List<TopicPartition> ownedPartitions = consumerEntry.getValue();
 
             List<TopicPartition> consumerAssignment = assignment.get(consumer);
-            int i = 0;
-            // assign the first N partitions up to the max quota, and mark the remaining as being revoked
-            for (TopicPartition tp : ownedPartitions) {
-                if (i < maxQuota) {
-                    consumerAssignment.add(tp);
-                    unassignedPartitions.remove(tp);
-                } else {
-                    allRevokedPartitions.add(tp);
-                }
-                ++i;
-            }
 
             if (ownedPartitions.size() < minQuota) {
+                // the expected assignment size is more than consumer have now, so keep all the owned partitions
+                // and put this member into unfilled member list
+                if (ownedPartitions.size() > 0) {
+                    consumerAssignment.addAll(ownedPartitions);
+                    assignedPartitions.addAll(ownedPartitions);
+                }
                 unfilledMembers.add(consumer);
+            } else if (ownedPartitions.size() >= maxQuota && numMaxCapacityMembers < numExpectedMaxCapacityMembers) {
+                // consumer owned the "maxQuota" of partitions or more, and we're still under the number of expected max capacity members
+                // so keep "maxQuota" of the owned partitions, and revoke the rest of the partitions
+                numMaxCapacityMembers++;
+                List<TopicPartition> maxQuotaPartitions = ownedPartitions.subList(0, maxQuota);
+                consumerAssignment.addAll(maxQuotaPartitions);
+                assignedPartitions.addAll(maxQuotaPartitions);
+                allRevokedPartitions.addAll(ownedPartitions.subList(maxQuota, ownedPartitions.size()));
             } else {
-                // It's possible for a consumer to be at both min and max capacity if minQuota == maxQuota
-                if (consumerAssignment.size() == minQuota)
-                    minCapacityMembers.add(consumer);
-                if (consumerAssignment.size() == maxQuota)
-                    maxCapacityMembers.add(consumer);
+                // consumer owned at least "minQuota" of partitions but we're already at the allowed number of max capacity members
```

Comment:
I wonder if we need to handle one more case? I'm thinking about where the consumer had exactly `minQuota` previously owned partitions but we're not yet at `numExpectedMaxCapacityMembers`. Should we consider the consumer to be "unfilled" since it may be that this consumer needs to eventually get to `maxQuota`?

I tried to think up an example where this current logic would break, and I couldn't. It's possible that there isn't one just by mathematical coincidence -- eg if there's just one consumer which had `minQuota` previously owned partitions then in theory it might need to have one more assigned, but in reality if there's only one consumer then min and max quota will always be equal. Just wondering if you could give this some thought and make sure we're not missing something here.

(Even if not, we should update the comment here since it's a bit misleading -- we don't _know_ that we're at the allowed max capacity members yet. Maybe we can instead leave a comment explaining why we think the above case is not possible, assuming you do come to that same conclusion)

Reply:
You found a potential bug here!!! You're awesome!!! It's great we found and fix it here, not releasing it to the users! So, again, it implied the tests didn't cover all cases! I'll add more tests for it. Thank you!

Reply:
I think this is actually possible even with more than one consumer, since in we can have both num.partitions AND num.consumers change. When that happens, relying on prev.assigned tasks would not guarantee that stickiness would be just sufficient to achieve the end goal.

Reply:
nit: also in practice, I think the common case would be `minQuota == maxQuota` and `numExpectedMaxCapacityMembers == 0`. So the name `numExpectedMaxCapacityMembers` here could be a bit misleading. Maybe rename to `expectedNumMembersHavingMorePartitions` and `numMembersHavingMorePartitions` instead?

Reply:
Yes, this could happen, and just updated. Also the renaming is better. Thanks.